Fix FESeries with non-primitive FEs.

adamqc opened the pull request:
Closes #11791.

Add a component parameter to FESeries::Fourier/Legendre to
make them working with non-primitive elements.

marcfehling commented on include/deal.II/fe/fe_series.h:
```diff
@@ -364,6 +381,12 @@ namespace FESeries
+     * Which component of FiniteElement should be used to calculate the
+     * expansion.
+     */
+    unsigned int component;
```
You can make this `const`, too.

bangerth replied:
```suggestion
    const unsigned int component;
```

marcfehling commented on source/fe/fe_series_fourier.cc:
```diff
@@ -179,16 +189,23 @@ namespace FESeries
              n_coefficients_per_direction.size() == q_collection.size(),
            ExcMessage("All parameters are supposed to have the same size."));
 
+#ifdef DEBUG
+    for (unsigned int i = 0; i < fe_collection.size(); ++i)
+      AssertIndexRange(component, fe_collection[i].n_components());
+#endif
```
Well placed assertion! :-)

bangerth replied:
But I think you don't actually need it in this generality. `hp::FECollection` enforces that all elements have the same number of vector components, so it is probably enough to just check for `[0]`.

marcfehling commented on include/deal.II/fe/fe_series.h:
```diff
@@ -96,24 +96,29 @@ namespace FESeries
             const hp::FECollection<dim, spacedim> &fe_collection,
-            const hp::QCollection<dim> &           q_collection);
+            const hp::QCollection<dim> &           q_collection,
+            const unsigned int                     component = 0);
```
Here and other occasions: I would suggest to use `numbers::invalid_unsigned_int` as a default parameter.

If users provide vector-valued FEs and use this function, this function will work only on the very first component -- unknowing to the user. Instead, with the invalid parameter, you can check in the constructor with an assertion if the provided `fe_collection` consists of scalar or vector-valued finite elements, and then set `component` to zero if it only contains scalar FEs. This way users will be alarmed whether they are doing something unintentionally.

bangerth replied:
Yes, I agree.

marcfehling commented on include/deal.II/fe/fe_series.h:
```diff
@@ -96,24 +96,29 @@ namespace FESeries
      * each direction, @p fe_collection is the hp::FECollection for which
      * expansion will be used and @p q_collection is the hp::QCollection used to
      * integrate the expansion for each FiniteElement in @p fe_collection.
+     * @p component defines which component of each FiniteElement will be used,
+     * it is mainly used for non-primitive elements.
```
Here and other occasions:
```suggestion
     *
     * As the Fourier expansion can only be performed on scalar fields, this class does not operate on vector-valued finite elements and will therefore throw an assertion. However, each component of a finite element field can be treated as a scalar field, respectively, on which Fourier expansions are again possible. For this purpose, the optional parameter @p component defines which component of each FiniteElement will be used.
```

marcfehling reviewed (changes requested):
Looks good so far!

Could you provide a test that provides the whole smoothness estimation workflow? You can basically take your minimal example of #11791 for this, and initialize your `solution` vector with meaningful values, if necessary.

I somehow doubt that the estimator functions in the `SmoothnessEstimator` namespace will work with non-primitive finite elements out of the box. See here and other occasions: https://github.com/dealii/dealii/blob/c9feb145b3d368600a145c094dc2739731ef1530/source/numerics/smoothness_estimator.cc#L131-L136

For non-primitive finite elements, you will not need all `dof_values` on the expansion, but only those associated with the `component` of the `fe_legendre` object.

adamqc commented:
> For non-primitive finite elements, you will not need all `dof_values` on the expansion, but only those associated with the `component` of the `fe_legendre` object.

Good point.
I'm not quite familiar with the `FiniteElement` class, how can I get the `dof_values` associate with the `component`?

marcfehling commented:
> I'm not quite familiar with the `FiniteElement` class, how can I get the `dof_values` associate with the `component`?

I haven't worked much with vector-valued problems yet, so maybe someone else has a better idea than me:

Maybe the module documentation [Handling vector valued problems - Extracting data from solutions](https://www.dealii.org/developer/doxygen/deal.II/group__vector__valued.html#VVExtracting) helps. The approach described there would involve using `FEValues` and `FEValuesExtractor`, but since we only need the correct indices from the finite element solution on each cell, I think there is a better approach than that.

[`step-8`](https://www.dealii.org/developer/doxygen/deal.II/step_8.html#ElasticProblemassemble_system) describes how to use [`FiniteElement::system_to_component_index()`](https://www.dealii.org/developer/doxygen/deal.II/classFiniteElement.html#a86644fe67824373cd51e9ff7fca94f8c). Maybe you can use the opposite function [`FiniteElement::component_to_system_index()`](https://www.dealii.org/developer/doxygen/deal.II/classFiniteElement.html#ae2ea16b60a6fc644a9bc7097703a53e8) to extract those dof values that belong to your component from [`DoFCellAccessor::get_dof_values()`](https://www.dealii.org/developer/doxygen/deal.II/classDoFCellAccessor.html#a1702d1e9c1cb4fb07b70c38ac9165aaa). It would be handy to introduce an overload of the latter function in this context.

bangerth commented on include/deal.II/fe/fe_series.h:
```diff
@@ -204,6 +209,12 @@ namespace FESeries
+     * Which component of FiniteElement should be used to calculate the
+     * expansion.
+     */
+    unsigned int component;
```
```suggestion
    const unsigned int component;
```

bangerth commented on include/deal.II/fe/fe_series.h:
```diff
@@ -96,24 +96,29 @@ namespace FESeries
             const hp::FECollection<dim, spacedim> &fe_collection,
-            const hp::QCollection<dim> &           q_collection);
+            const hp::QCollection<dim> &           q_collection,
+            const unsigned int                     component = 0);
```
This constructor is deprecated anyway, you probably don't need to add the `component` argument.

bangerth commented on include/deal.II/numerics/smoothness_estimator.h:
```diff
@@ -232,10 +232,13 @@ namespace SmoothnessEstimator
     FESeries::Legendre<dim, spacedim>
-    default_fe_series(const hp::FECollection<dim, spacedim> &fe_collection);
+    default_fe_series(const hp::FECollection<dim, spacedim> &fe_collection,
+                      const unsigned int                     component = 0);
```
same here with the `numbers::invalid_unsigned_int`.

bangerth commented on source/fe/fe_series_fourier.cc:
```diff
@@ -65,14 +65,16 @@ namespace
         sum += std::exp(std::complex<double>(0, 1) * (k_vector * x_q)) *
-               fe.shape_value(j, x_q) * quadrature.weight(q);
+               fe.shape_value_component(j, x_q, component) *
+               quadrature.weight(q);
```
Yes, I think this is really the only necessary change, at least for primitive vector-valued elements

bangerth commented on source/fe/fe_series_fourier.cc:
```diff
@@ -179,16 +189,23 @@ namespace FESeries
     , fe_collection(&fe_collection)
     , q_collection(q_collection)
     , fourier_transform_matrices(fe_collection.size())
+    , component(component)
```
With the invalid value, this would
```suggestion
    , component(component != numbers::invalid_unsigned_int ? component : 0)
```
and then you would add the following assertion below:
```
  if (fe_collection.n_component() > 0)
    Assert (component != numbers::invalid_unsigned_int,
                ExcMessage ("For vector-valued problems, you need to explicitly specify for which vector "
                        "component you will want to do a Fourier decomposition by setting the 'component' "
                        "argument of this constructor.")
```

marcfehling replied:
It would make sense to give the constructor parameter and the class member different names to avoid ambiguity, e.g., `component_` for the former and `component` for the latter.

adamqc commented:
I have addressed all your comments except @marcfehling 's comment about `dof_values`. I think that for non-primitive elements such as Nedelec and RT that all components are coupled with all dofs, they can not be separated simply. Am I right?

masterleinad commented:
> I think that for non-primitive elements such as Nedelec and RT that all components are coupled with all dofs, they can not be separated simply. Am I right?

`FEValuesBase::shape_value_component` also works for non-primitive elements. The question is rather if it makes sense to only compute a Fourier/Legendre expansion corresponding to one component in this case but the implementation would still allow this AFAICT.

bangerth commented:
That is the same as for `KellyErrorEstimator`, by the way. If that class does it per component, then so should this one here.

marcfehling commented on tests/numerics/smoothness_estimator_03.output:
```diff
@@ -0,0 +1,3 @@
+
+DEAL::Ok
+DEAL::Ok
```
Judging from the source code, this should be in capitals. Did you run this test yet and checked if it passes?

```suggestion
DEAL::OK
DEAL::OK
```

adamqc replied:
Sorry. The source code was modified unintentionally.

marcfehling commented on tests/numerics/smoothness_estimator_03.cc:
```diff
@@ -0,0 +1,72 @@
+
+#include "../tests.h"
+
+// This test is used to make sure FESeries::Fourier/Legendre
+// woring with non-primitive elements
```
Would you move your brief test description before the `include` statements, and add the deal.II copyright preamble in the front of this file?

You could draw inspiration from the other smoothness_estimator tests.

```suggestion
// This test is used to make sure that FESeries::Fourier/Legendre
// work with non-primitive elements
```

marcfehling reviewed (changes requested):
Minor comments and a few very pedantic suggestions.

bangerth commented:
/rebuild

bangerth commented on include/deal.II/fe/fe_series.h:
```diff
@@ -96,10 +97,18 @@ namespace FESeries
             const hp::FECollection<dim, spacedim> &fe_collection,
-            const hp::QCollection<dim> &           q_collection);
+            const hp::QCollection<dim> &           q_collection,
+            const unsigned int component = numbers::invalid_unsigned_int);
```
Can you document how the default value is interpreted? As in, the default value can only be used for a scalar finite element, in which case it indicates that the sole component is to be decomposed. For vector-valued elements, a non-default value of that argument has to be explicitly provided.

bangerth commented on include/deal.II/fe/fe_series.h:
```diff
@@ -262,10 +278,18 @@ namespace FESeries
              const hp::FECollection<dim, spacedim> &fe_collection,
-             const hp::QCollection<dim> &           q_collection);
+             const hp::QCollection<dim> &           q_collection,
+             const unsigned int component = numbers::invalid_unsigned_int);
```
same here

bangerth commented on include/deal.II/numerics/smoothness_estimator.h:
```diff
@@ -232,10 +232,19 @@ namespace SmoothnessEstimator
-    default_fe_series(const hp::FECollection<dim, spacedim> &fe_collection);
+    default_fe_series(
+      const hp::FECollection<dim, spacedim> &fe_collection,
+      const unsigned int component = numbers::invalid_unsigned_int);
```
and here

bangerth commented on include/deal.II/numerics/smoothness_estimator.h:
```diff
@@ -458,10 +467,19 @@ namespace SmoothnessEstimator
-    default_fe_series(const hp::FECollection<dim, spacedim> &fe_collection);
+    default_fe_series(
+      const hp::FECollection<dim, spacedim> &fe_collection,
+      const unsigned int component = numbers::invalid_unsigned_int);
```
and here

adamqc replied:
I have addressed your comments about the documentation.

bangerth reviewed (commented):
Excellent! I have just a couple of comments.

But here is one other question. At the end of the `calculate()` function, we have
```
    const FullMatrix<CoefficientType> &matrix =
      fourier_transform_matrices[cell_active_fe_index];

    [...]

    for (unsigned int i = 0; i < unrolled_coefficients.size(); i++)
      for (unsigned int j = 0; j < local_dof_values.size(); j++)
        unrolled_coefficients[i] += matrix[i][j] * local_dof_values[j];

    fourier_coefficients.fill(unrolled_coefficients.begin());
```
That last block computes the expansion by multiplying the matrix that contains the expansion of each shape function (i.e., the component of interest of each shape function) by the DoF values on each cell.

Many of the columns of this matrix will be zero, except for those columns for which a shape function is nonzero in the selected component. I think that means that all DoF values for other vector components simply don't make a difference, as indeed should be the case. So the fact that we built a potentially much larger matrix than necessary doesn't matter, right?

One could, in a follow-up, optimize the integration of these matrices from the current code
```
        for (unsigned int k = 0; k < n_coefficients_per_direction[fe]; ++k)
          for (unsigned int j = 0; j < fe_collection[fe].n_dofs_per_cell(); ++j)
            fourier_transform_matrices[fe](k, j) =
              integrate(fe_collection[fe], q_collection[fe], k_vectors(k), j);
```
to one that switches the order of loops and simply ignores those `j` indices for which we know that the shape function is zero in the selected component. But let's do that in a later patch.

adamqc commented:
> One could, in a follow-up, optimize the integration of these matrices from the current code
> to one that switches the order of loops and simply ignores those `j` indices for which we know that the shape function is zero in the selected component. But let's do that in a later patch.

I will investigate how to achieve this.

bangerth reviewed (approved):
Great, thanks a lot!
